### PR TITLE
fix: Move snapd sync to custom post-refresh hook

### PR DESCRIPTION
**Pull request by bschimke95**
## Description

We disable the snapd/k8sd sync on upgrades because k8sd is not available at that time. Right now, we re-enable the sync in the snap `post-refresh` hook
The snapd/k8sd sync requires access to k8sd to obtain/sync the values. The snap `post-refresh` hook [is executed before the services started](https://snapcraft.io/docs/supported-snap-hooks#heading--post-refresh) and therefore wait until timeout.

## Solution

This commit introduces a custom post-refresh hook that runs after k8sd started and moves the config sync into that hook. In the future, other machinery also depends on k8sd after a refresh (e.g. feature upgrades) so it makes sense to have this kind of common entry point to extend later.

I verified the upgrade from rev 2500 to this version (`latest/edge/ben-refresh-hook`) works for a three node cluster. Once this fix is merged, we can also re-enable the upgrade/downgrade tests until latest/edge.

image

## Issue
-

## Backport

The post-refresh mechanism was [backported to ](https://github.com/canonical/k8s-snap/pull/1164) 1.32, so we also need to backport this fix.

## Checklist

- [x] PR title formatted as `type: title`
- [ ] Covered by unit tests - N/A
- [ ] Covered by integration tests - already covered by upgrade tests
- [ ] Documentation updated - N/A
- [x] CLA signed
- [x] Backport label added if necessary 

If any item on the checklist is not complete, please provide justification why.

### CODE REVIEW

**Review by berkayoz — APPROVED** [review]
LGTM. Lock files are not the cleanest approach however I don't see an easier way for now. This should get us going and we can revisit later.

File: src/k8s/pkg/k8sd/app/hooks_start.go [hunk 1]
```diff
@@ -24,6 +25,20 @@ func (a *App) onStart(ctx context.Context, s state.State) error {
 		}
 	}()
 
+	// Check if a refresh was performed and if so, run the custom post-refresh hook
+	isPostRefresh, err := utils.FileExists(a.snap.PostRefreshLockPath())
```
Comment (addyess) [inline src/k8s/pkg/k8sd/app/hooks_start.go:29]: what kind of errors can `utils.FileExists` encounter? Permissions to look at a path?  Encountering something that isn't path-like?

File: src/k8s/pkg/k8sd/app/hooks_start.go [hunk 2]
```diff
@@ -24,6 +25,20 @@ func (a *App) onStart(ctx context.Context, s state.State) error {
...
+	if err != nil {
+		return fmt.Errorf("failed to check if snap is post-refresh: %w", err)
+	}
+	if isPostRefresh {
+		if err := a.postRefreshHook(ctx, s); err != nil {
+			return fmt.Errorf("failed to run post-refresh hook: %w", err)
+		}
```
Comment (addyess) [inline src/k8s/pkg/k8sd/app/hooks_start.go:36]: So we will fail in k8sd -- but because it didn't fail in the post-refresh hook -- what will snapd do?  It can't rollback the snap anymore?  Are we just left with a k8ds that's on the new revision but isn't rolledback?

File: src/k8s/pkg/k8sd/app/hooks_start.go [hunk 3]
```diff
@@ -24,6 +25,20 @@ func (a *App) onStart(ctx context.Context, s state.State) error {
...
+		if err := os.Remove(a.snap.PostRefreshLockPath()); err != nil {
+			return fmt.Errorf("failed to remove post-refresh lock file: %w", err)
```
Comment (addyess) [inline src/k8s/pkg/k8sd/app/hooks_start.go:38]: This feels like a super ugly hole to land in.  We've upgraded, we've run the the `postRefreshHook`, but k8sd can't start b/c it can't delete the lock file.

**Review by addyess — COMMENTED** [review]
Mostly questions regarding the now non-snapd-rollback on genuinely failed refreshes that could have been detected in post-refresh but now won't be.

File: src/k8s/pkg/k8sd/app/hooks_start.go (hunk 1 above)
Comment (bschimke95) [reply src/k8s/pkg/k8sd/app/hooks_start.go:29]: It uses os.Stat under the hood so basically every I/O syscall issue can happen, e.g.:
os.ErrNotExist – File does not exist
os.ErrPermission – Permission denied
syscall.EINVAL – Invalid file name or argument
syscall.ELOOP – Too many symbolic links encountered
syscall.EIO – I/O error (disk or hardware failure)
syscall.ENOSPC – No space left on device (rare for os.Stat)
syscall.ENAMETOOLONG – File name too long
syscall.EROFS – Read-only file system
syscall.ENOTDIR – A component of the path is not a directory
syscall.EFAULT – Invalid memory address (rare case)

File: src/k8s/pkg/k8sd/app/hooks_start.go (hunk 3 above)
Comment (bschimke95) [reply src/k8s/pkg/k8sd/app/hooks_start.go:38]: Well, yes but also no. Removing a file should just work in 99% of the cases or be transient. Worst case, the user has to restart the node. Then, the post-refresh will kick in again but it should be mostly (?!) idempotent. And then the lock removal would be retried.

Comment (addyess) [reply src/k8s/pkg/k8sd/app/hooks_start.go:38]: Good enough for me

File: src/k8s/pkg/k8sd/app/hooks_start.go (hunk 1 above)
Comment (addyess) [reply src/k8s/pkg/k8sd/app/hooks_start.go:29]: ack -- a whole bunch of icky things that neither snapd rollback nor k8ds can manage resolving.

File: src/k8s/pkg/k8sd/app/hooks_start.go (hunk 2 above)
Comment (bschimke95) [reply src/k8s/pkg/k8sd/app/hooks_start.go:36]: That is a good point. Yes, we basically loose snapds roll-back capabilities but I feel like this is what we signed up for if we tie k8sd configuration to snap configuration. 
Eventually, we'll track an upgrade in a CRD (#1207). If we have that, then the node would mark the upgrade as failed if the post-refresh hook fails and the orchestration tool would be in charge of the rollback. Which IMHO makes more sense anyway.

Comment (addyess) [reply src/k8s/pkg/k8sd/app/hooks_start.go:36]: Well lovely.  I think owning our own space rather than relying on snapd to "rollback" is a better solution to.  At the moment it LGTM, i just needed to make sure I was understanding what you're selling here.
